docs: add comprehensive generalized wrappers documentation (#595)
cowprotocol/docs: docs/cow-protocol/integrate/wrappers.mdx

# Wrapper Integration

Generalized wrappers enable custom logic to execute surrounding CoW Protocol settlements. This guide covers both using existing wrappers in your orders and building new wrapper contracts.

**Choose your path:**
- **[Building New Wrappers](#for-wrapper-developers)** - Develop custom wrapper contracts for advanced DeFi workflows
- **[Using Existing Wrappers](#for-order-creators-and-frontend-developers)** - Add wrappers to your orders and integrate into your application
- **[Executing Wrappers as a Solver](#for-solvers)** - Everything solvers need to know to execute wrapper orders

---

## For Wrapper Developers

This section covers building new wrapper contracts from scratch.

### Overview

To build a wrapper, you will:

1. Inherit from the `CowWrapper` abstract contract
2. Implement `_wrap()` with your pre/post-settlement logic
3. Implement `validateWrapperData()` for input validation
4. Test thoroughly on testnets
5. Review the [Implementation Requirements](../reference/contracts/periphery/wrapper.mdx#implementation-requirements-for-integrators) to ensure your wrapper meets the acceptance requirements
6. Submit for allowlist approval by the CoW Protocol team

> 

Reviewer: Let's stress that it's difficult to build one. We want to prevent people to expend effort for nothing and we want to avoid spam submissions.

```suggestion

Building a wrapper is a significant undertaking.
An insecure wrapper would put user and protocol funds at risk.
Make sure you understand CoW Protocol well before starting.
A contract audit by a reputable firm is mandatory before a wrapper is considered for approval.
```

The wrappers execute in sequence: Wrapper1 → Wrapper2 → Settlement → Wrapper2 (post) → Wrapper1 (post). Note that wrappers from other users' orders may be interspersed, though this should generally not affect the execution of the order.

### Validating Wrapper Configuration

Before submitting an order, you can confirm the wrapper encoding is valid using [`CowWrapperHelper`](https://gist.github.com/kaze-cow/8c8a49f3305519948b1ec868c7efa5b9):

```typescript
// On-chain validation (via Ethers/Viem)
const helper = new ethers.Contract(HELPER_ADDRESS, HELPER_ABI, provider)

const isValid = await helper.verifyAndBuildWrapperData([
    {
        target: '0xSomeAddress',
        data: '0xSomeData'
    },
    {
        target: '0xSomeAddress2',
        data: '0x'
    }
])
// Returns encoded chainedWrapperData if valid, reverts if invalid
```

This checks:
- All wrappers are allowlisted
- Each wrapper can parse its data successfully

The CoW Wrapper Helpers contract is deployed via `CREATE2`, and is being rolled out to the following addresses on each network supported by CoW:
* Production: `0x8C8f88F3081A7d54fCbC2Bd7C6Cf5E9Fc26e30E9`
* Staging: `0xfd1a35C8E7AEC58a4bB6AeDca0B2Bf9c9Be9E07F`

> 

Reviewer: Can you move this to [this page](https://docs.cow.fi/cow-protocol/reference/contracts/periphery) and add a link here for consistency?
Eventually I hope we change how we generate that page thanks to the new deployment repo!

---

## For Solvers

This section explains how to execute settlements that include wrapper contracts as part of your solver implementation.

### Detecting Wrapper Orders

Wrappers are specified in the order's [`appData`](../reference/core/auctions/schema.md) under the `wrappers` field:

```json
{
  "wrappers": [
    {
      "target": "0x1234...",
      "data": "0xabcdef...",
      "isOmittable": false
    }
  ]
}
```

**Fields:**
- **`target`**: Address of the wrapper contract (treat as opaque - don't parse)
- **`data`**: Wrapper-specific data (treat as opaque - don't parse)

> - **`isOmittable`**: Whether the `target` and `data` must be executed unmodified as part of the order execution

Reviewer: ([Continuing](https://github.com/cowprotocol/docs/pull/595#discussion_r2922649447).)
I see "parsing" as the process of interpreting the opaque string to the expected format (i.e., one Ethereum address and one bytestring). Not parsing means that solvers must literally accept `"target": "invalid address"`. The cost of doing this is negligible.

> Also regarding target not being an address, [the appData schema](https://github.com/cowprotocol/cow-sdk/blob/main/packages/app-data/src/schemas/wrappers/v0.2.0.json#L18) literally constrains it to be a [hex with 40 characters](https://github.com/cowprotocol/cow-sdk/blob/main/packages/app-data/src/schemas/definitions.json#L16) (ignoring checksumming). [Same happens in the autopilot and api](https://github.com/cowprotocol/services/blob/main/crates/app-data/src/app_data.rs#L71) when they receive app data.

Users can sign an order with nonsense app data (at least, this was the case last time I checked, through the ETH flow). Anyway, even if the orderbook rejects it, that's still a valid order.
We don't provide any guarantee that the app data is valid, since it's user controlled. We don't want to have the orderbook guarantee the validity of the app data. **The orderbook is not a gatekeeper for the protocol.** We don't want it to become one. Points of centralization should be avoided and this assumption would be a small step towards centralization.

Also, we gain really nothing out of this. What's the point of saving an address check and a bytes decoding for the solver?

What I see being the point here is that we don't want solvers to "decode" the data, the mechanism is the same for all wrappers.

```suggestion
- **`target`**: Address of the wrapper contract
- **`data`**: Wrapper-specific data
- **`isOmittable`**: Whether the `target` and `data` must be executed unmodified as part of the order execution

`target` may be any address and `data` is arbitrary bytes that will be passed to the target. There is generally no need for decoding the wrapper data since the execution process uses the same opaque format on each wrapper.
```

# Wrapper Integration

Generalized wrappers enable custom logic to execute surrounding CoW Protocol settlements. This guide covers both using existing wrappers in your orders and building new wrapper contracts.

**Choose your path:**
- **[Building New Wrappers](#for-wrapper-developers)** - Develop custom wrapper contracts for advanced DeFi workflows
- **[Using Existing Wrappers](#for-order-creators-and-frontend-developers)** - Add wrappers to your orders and integrate into your application
- **[Executing Wrappers as a Solver](#for-solvers)** - Everything solvers need to know to execute wrapper orders

---

## For Wrapper Developers

This section covers building new wrapper contracts from scratch.

### Overview

To build a wrapper, you will:

1. Inherit from the `CowWrapper` abstract contract
2. Implement `_wrap()` with your pre/post-settlement logic
3. Implement `parseWrapperData()` for input validation
4. Test thoroughly on testnets

> 5. Submit for allowlist approval by the CoW Protocol team

Reviewer: I'm worried about this, the last thing we want to do is wading through a bunch of half-assed wrapper implementations. We should really add some details on this, the bare minimum is a contract audit from a well-known provider.

Author: I went ahead and put in our full list of implementation requirements and a link directly to it here 8efd607

> Read up on the details of creating a wrapper from [the wrapper contract information page](../reference/contracts/periphery/wrapper.mdx).

Reviewer 2: Read up on the details of creating a wrapper from [the wrapper --> Read more about the details of creating a wrapper from [the wrapper

Author: 5563ab8

---

## For Order Creators and Frontend Developers

This section shows you how to use existing wrappers in your orders and integrate wrapper support into your application.

### Adding Wrappers to Orders

To use a wrapper in your order, add it to the `appData` when creating the order:

### AppData Structure

```typescript
interface WrapperCall {
    target: string;        // Wrapper contract address (must be allowlisted)
    data?: string;         // Optional hex-encoded wrapper-specific data

>     is_omittable?: boolean; // Whether solver can skip this wrapper (default: false)

Reviewer: I'm not up to speed with this. Is this how it works now? I thought we decided to not use `is_omittable` in the end.
Also, is it right that it's snake_case and not camelCase?

Author: >I'm not up to speed with this. Is this how it works now? I thought we decided to not use is_omittable in the end.

It is how it works, yes, and it is upstreamed in both the [`services`](https://github.com/cowprotocol/services/blob/82d3e7c5d72410f3438009a7a595c21f25382395/crates/app-data/src/app_data.rs#L78) and [`cow-sdk`](https://github.com/cowprotocol/cow-sdk/blob/main/packages/app-data/src/schemas/wrappers/v0.2.0.json#L25) repos.

Its snake case in rust due to rust code style, and camel case everywhere else. The integration guide should definitely be displaying the camelCase 99f2fc9

    // ... standard order fields (sellToken, buyToken, amounts, etc.)

    appData: {
        // ... other appData fields
        wrappers: [
            {
                target: "0x1234...",     // Flash loan wrapper address
                data: "0xabcd...",       // Encoded flash loan params
                is_omittable: false      // Must execute
            }
        ]
    }
}

// Submit order
const orderId = await orderBookApi.sendOrder(orderCreation)
```

### Example: Multiple Wrappers (Nested)

You can chain multiple wrappers in a single order:

```typescript
appData: {
    wrappers: [
        {
            target: "0x1111...",     // Flash loan wrapper
            data: "0xaaaa...",
            is_omittable: false
        },
        {
            target: "0x2222...",     // Leverage wrapper
            data: "0xbbbb...",
            is_omittable: false
        }
    ]
}
```

The wrappers execute in sequence: Wrapper1 → Wrapper2 → Settlement → Wrapper2 (post) → Wrapper1 (post). Note that wrappers from other users' orders may be interspersed, though this should generally not affect the execution of the order.

### Validating Wrapper Configuration

> Before submitting an order, you can confirm the wrapper encoding is valid using `CowWrapperHelper`:

Reviewer: Here it needs a link to the helper address, or some way to recover it.

Author: yea true. We haven't officially deployed this contract yet (my fault, sorry), so for right now [I put it in a gist](https://gist.github.com/kaze-cow/8c8a49f3305519948b1ec868c7efa5b9) and linked it here. We will need to come back here and update it when it becomes an official cow protocol deployment (probably in the next few weeks). e6d4223

```typescript
// On-chain validation (via Ethers/Viem)
const helper = new ethers.Contract(HELPER_ADDRESS, HELPER_ABI, provider)

const isValid = await helper.verifyAndBuildWrapperData(
    ["0x1234...", "0x5678..."],  // Wrapper addresses
    ["0xabcd...", "0xef01..."]   // Wrapper data
)
// Returns encoded wrapperData if valid, reverts if invalid
```

This checks:
- All wrappers are allowlisted
- All wrappers use the same settlement contract
- Each wrapper can parse its data successfully

---

## For Solvers

This section explains how to execute settlements that include wrapper contracts as part of your solver implementation.

## Detecting Wrapper Orders

> Wrappers are specified in the order's [`appData`](../reference/core/auctions/schema.md) under the `wrappers` field:

Reviewer 2: order's ---> orders

Author: I believe the grammar is it is now is correct here. The appData is "possessed" by the order.

```json
{
  "wrappers": [
    {
      "target": "0x1234...",
      "data": "0xabcdef...",
      "is_omittable": false
    }
  ]
}
```

**Fields:**
- **`target`**: Address of the wrapper contract (treat as opaque - don't parse)
- **`data`**: Wrapper-specific data (treat as opaque - don't parse)
- **`is_omittable`**: Critical flag for solver requirements (see below)

:::warning
Wrapper addresses and data should be treated as opaque values. Do not attempt to parse or validate them - simply pass them through in your encoding.
:::

## Solver Requirements

### 1. Execute Non-Omittable Wrappers

Orders with `"is_omittable": false` **MUST** be executed with the specified wrapper. You may be slashed for not doing so.

If `"is_omittable": true`, you MAY skip the wrapper if you find a better solution without it.

### 2. Verify Wrapper Authentication

All approved wrappers will be approved by the DAO and registered in [`GPv2AllowlistAuthenticator`](../reference/contracts/core/allowlist.md). It is recommended to verify wrappers are allowlisted before including them in settlements.

### 3. Include Wrappers in Solution

> Whether using the CoW-provided driver or your own, ensure wrapper information flows from the auction to your final settlement transaction.

Reviewer 2: CoW-provided driver or your own --> CoW can't be on its own, needs to be CoW DAO or CoW Protocol

Author: ad8f9e6

## Encoding Wrapper Settlements

:::info

> If you're using the driver provided by the CoW team, you can skip this section - the driver handles encoding automatically.

Reviewer 2: As above no single use of CoW

Author: 3e58601

    }

    return wrapperData;
}
```

**Encoding Structure:**
```
For 2 wrappers:
[length₁(2B)][data₁][address₂(20B)][length₂(2B)][data₂]

For 3 wrappers:
[length₁(2B)][data₁][address₂(20B)][length₂(2B)][data₂][address₃(20B)][length₃(2B)][data₃]
```

#### 4. Call wrappedSettle

```typescript
const tx = {
    to: wrappers[0].target,
    data: wrapperContract.interface.encodeFunctionData("wrappedSettle", [
        settleData,      // From step 2
        chainedWrapperData      // From step 3
    ])
}
```

### Using CowWrapperHelper

Alternatively, use the `CowWrapperHelper` contract for encoding and validation:

```solidity
// On-chain encoding and validation
address[] memory wrapperAddresses = [wrapper1, wrapper2];
bytes[] memory wrapperDatas = [data1, data2];

bytes memory wrapperData = CowWrapperHelper(HELPER).verifyAndBuildWrapperData(
    wrapperAddresses,
    wrapperDatas
);
```

In addition to producing the correct encoded data, the helper also validates:
- All wrappers are allowlisted
- All wrappers use the same settlement contract
- Each wrapper can parse its data

## Accumulating Wrappers in Solutions

### Using CoW-Provided Driver

> If you're using the CoW team's driver, wrappers are handled automatically. Just ensure your solver process includes wrapper information in the solution output.

Reviewer 2: as above, no single 'CoW'

Author: 3e58601

#### Custom Implementation

If implementing your own solver:

1. **Collect wrappers from orders**: As you process orders in an auction, collect all `wrappers` arrays from order appData
2. **Aggregate for batch**: Combine wrappers needed for all orders in the settlement batch
3. **Include in solution**: Add aggregated wrappers to your solution structure
4. **Encode transaction**: Use the encoding algorithm above to construct the final transaction

**Example (conceptual):**

```typescript
// Process auction
const solution = {
    orders: [...],
    wrappers: []  // Collect here
};

for (const order of auction.orders) {
    if (order.appData.wrappers) {
        // Add wrappers needed for this order
        solution.wrappers.push(...order.appData.wrappers);
    }
}

// Encode final transaction
const tx = encodeWrapperSettlement(solution);

> ```

Reviewer: All the magic is hidden in `encodeWrapperSettlement`. I think it's fair to assume that the solvers know how to create the settle calldata and use that together with `verifyAndBuildWrapperData`.

Author: definitely some clariifcation was needed here. Referred explicitly to the encoding section above and also showed the actual construction of the txn object dc0bf1d

# Wrapper Integration

Generalized wrappers enable custom logic to execute surrounding CoW Protocol settlements. This guide covers both using existing wrappers in your orders and building new wrapper contracts.

**Choose your path:**
- **[Building New Wrappers](#for-wrapper-developers)** - Develop custom wrapper contracts for advanced DeFi workflows
- **[Using Existing Wrappers](#for-order-creators-and-frontend-developers)** - Add wrappers to your orders and integrate into your application
- **[Executing Wrappers as a Solver](#for-solvers)** - Everything solvers need to know to execute wrapper orders

---

## For Wrapper Developers

This section covers building new wrapper contracts from scratch.

### Overview

To build a wrapper, you will:

1. Inherit from the `CowWrapper` abstract contract
2. Implement `_wrap()` with your pre/post-settlement logic

> 3. Implement `parseWrapperData()` for input validation

Reviewer: ```suggestion
3. Implement `validateWrapperData()` for input validation
```

```

### Example: Multiple Wrappers (Nested)

You can chain multiple wrappers in a single order:

```typescript
appData: {
    wrappers: [
        {
            target: "0x1111...",     // Flash loan wrapper
            data: "0xaaaa...",
            isOmittable: false
        },
        {
            target: "0x2222...",     // Leverage wrapper
            data: "0xbbbb...",
            isOmittable: false
        }
    ]
}
```

The wrappers execute in sequence: Wrapper1 → Wrapper2 → Settlement → Wrapper2 (post) → Wrapper1 (post). Note that wrappers from other users' orders may be interspersed, though this should generally not affect the execution of the order.

### Validating Wrapper Configuration

Before submitting an order, you can confirm the wrapper encoding is valid using [`CowWrapperHelper`](https://gist.github.com/kaze-cow/8c8a49f3305519948b1ec868c7efa5b9):

```typescript
// On-chain validation (via Ethers/Viem)
const helper = new ethers.Contract(HELPER_ADDRESS, HELPER_ABI, provider)

const isValid = await helper.verifyAndBuildWrapperData(
    ["0x1234...", "0x5678..."],  // Wrapper addresses
    ["0xabcd...", "0xef01..."]   // Wrapper data

> )

Reviewer: Inconsistent with the actual function:

```
    function verifyAndBuildWrapperData(WrapperCall[] memory wrapperCalls)
```

Author: consistantified 42ac22b

// Returns encoded wrapperData if valid, reverts if invalid
```

This checks:
- All wrappers are allowlisted

> - All wrappers use the same settlement contract

Reviewer: This isn't the case.

```suggestion
```

- Each wrapper can parse its data successfully

> Note: an official deployment of the `CowWrapperHelper` contract does not currently exist. If you would like to use this contract, please deploy it yourself using the [code in the gist](https://gist.github.com/kaze-cow/8c8a49f3305519948b1ec868c7efa5b9).

Reviewer: Not needed for the docs PR, but why not? We could easily deploy that.

Author: I am adding the deployment of this to the Cannon package and put the addresses that CREATE2 generates in :+1: 42ac22b  https://github.com/cowprotocol/euler-integration-contracts/pull/25

---

## For Solvers

This section explains how to execute settlements that include wrapper contracts as part of your solver implementation.

### Detecting Wrapper Orders

Wrappers are specified in the order's [`appData`](../reference/core/auctions/schema.md) under the `wrappers` field:

```json
{
  "wrappers": [
    {
      "target": "0x1234...",
      "data": "0xabcdef...",
      "isOmittable": false
    }
  ]
}
```

**Fields:**
- **`target`**: Address of the wrapper contract (treat as opaque - don't parse)
- **`data`**: Wrapper-specific data (treat as opaque - don't parse)
- **`isOmittable`**: Critical flag for solver requirements (see below)

:::warning
Wrapper addresses and data should be treated as opaque values. Do not attempt to parse or validate them - simply pass them through in your encoding.

> :::

Reviewer: Why are we so strict about this?
There is no real drawback in actually trying to parse if a solver wants to. They might have a good reason for it.
They kind of have to anyway, for example if `target` isn't an address, what should they do? What if the calldata is literally `0xinvalid`?
I think we should just remove all of this and just link to the section `### AppData Structure`.

Author: I suppose this could be reworded, but one common confusion for solvers is that they think they need to understand what is going on and how they should set up the actual solution. There have been multiple solvers during implementation that have asked me how they are supposed to parse/understand this data, and I have had to clarify that it is intentionally designed to be *not necessary* to do this. So this comment is designed to explicitly disperse those types of confusions. From a security perspective, yes, it seems like there is some use in that, but if solvers have to be so defensive about the use of wrappers, we have an unsurmountable problem in the design of wrappers in and of themselves. 

If you ask me, I think it would be more effective for a solver to **simulate the call** that would be executed and parse the simulation to see how it behaves (ex. what addresses it touches, funds transfers, etc.). Attempting to parse the wrapper data is kind of a losing game because we want to be able to add more new wrappers at any time and expecting solvers to maintain their own allowlist of solvers is going to make it much harder for us to introduce new wrappers in the future.

Also regarding `target` not being an address, [the `appData` schema](https://github.com/cowprotocol/cow-sdk/blob/main/packages/app-data/src/schemas/wrappers/v0.2.0.json#L18) literally constrains it to be a [hex with 40 characters](https://github.com/cowprotocol/cow-sdk/blob/main/packages/app-data/src/schemas/definitions.json#L16) (ignoring checksumming). [Same happens in the autopilot and api](https://github.com/cowprotocol/services/blob/main/crates/app-data/src/app_data.rs#L71) when they receive app data.

Author: after private discussion we decided to indeed reword this and to make sure we express that outside of basic validation, deeply understanding the data in the wrapper is not necessary, but we recommend solvers take measures to ensure security 42ac22b

### Solver Requirements

#### 1. Execute Non-Omittable Wrappers

Orders with `"isOmittable": false` **MUST** be executed with the specified wrapper. You may be slashed for not doing so.

If `"isOmittable": true`, you MAY skip the wrapper if you find a better solution without it.

#### 2. Verify Wrapper Authentication

> All approved wrappers will be approved by the DAO and registered in [`GPv2AllowlistAuthenticator`](../reference/contracts/core/allowlist.md). It is recommended to verify wrappers are allowlisted before including them in settlements.

Reviewer: Here instead it would be good to be strict. We're letting users tell the solver what to call and solvers should be aware of this.

```suggestion
All approved wrappers will be approved by the DAO and registered in [`GPv2AllowlistAuthenticator`](../reference/contracts/core/allowlist.md).

:::warning
The wrapper will be _directly_ called from the solver address and will mediate the settlement.
This value is provided _directly_ by the user and it may be malicious.
:::
```

> 

Reviewer: I think there should be an extra step here: decide whether to include the order in the settlement or not.
Since wrappers are likely to have complex behaviors, they would have to assess the revert risk based on the concrete wrapper implementation and may decide to skip it or adjust the scoring of the settlement accordingly.

Author: Do we have any specific advice on how to actually do that? I don't, other than simulating the order (which we already do in the driver and sometimes autopilot...). Seems like something we have to discuss. See previous comment about adding new wrappers and lack of solver involvement.

Author: Added additional docs here advising to simulate the transaction and perform any checks that may be necessary, especially confirming the transaction reverts or not 42ac22b

Reviewer: > which we already do in the driver and sometimes autopilot

The solver doesn't have to use the official driver. The driver does some checks, but solvers should know which of them are needed. A solver may be using the official driver, but it doesn't have to and the docs should reflect that.
For the autopilot, "sometimes" isn't enough, and even if it was "always," that's still not ideal because we don't want to add more responsibilities to the autopilot. We shouldn't use the autopilot as a gatekeeper (similar discussion [here](https://github.com/cowprotocol/docs/pull/595#discussion_r2930271284)).

### Encoding Wrapper Settlements

:::info
If you're using the driver provided by CoW Protocol, you can skip this section - the driver handles encoding automatically.
:::

#### Manual Encoding

To execute a settlement with wrappers:

##### 1. Change Transaction Target

Your transaction should call `wrappedSettle()` on the **first** wrapper (not the settlement contract):

```typescript
// Without wrapper
tx = {
    to: SETTLEMENT_CONTRACT,
    data: encodeSettle(...)
}

// With wrapper
tx = {
    to: wrappers[0].target,  // First wrapper address
    data: encodeWrappedSettle(...) // See below
}
```

##### 2. Construct settleData Parameter

The `settleData` parameter is the **exact** calldata you would send to `GPv2Settlement.settle()`, including the 4-byte function selector:

```typescript
const settleData = settlementContract.interface.encodeFunctionData("settle", [
    tokens,
    clearingPrices,
    trades,
    interactions
]);
// This is your settleData - unchanged from normal settlement
```

##### 3. Encode wrapperData Parameter

The `wrapperData` combines individual wrapper data with wrapper addresses:

```typescript

> function encodeWrapperData(wrappers: WrapperCall[]): string {

Reviewer: Can we link to a reference here instead of adding code?
Something I'm worried about is that there's no input validation whatsoever (we actually say "WARNING: don't validate"), then this would introduce super weird behaviors like a user injecting data in the `target` field and causing inconsistencies like being able to inject other wrappers or calls that aren't part of the expected execution.
We're also making a lot of assumptions, like the data being `0x` prefixed. Is it ok if the solvers skips the first byte provided by the user?
I'd say the user data should actually be validated at some point: it should conform to the app data schema, `target` should be an address, and `data` should be correctly encoded bytes, all `0x`-prefixed.

Author: > We're also making a lot of assumptions, like the data being 0x prefixed. Is it ok if the solvers skips the first byte provided by the user?

By the time the solver receives the data it will have been parsed into bytes by the autopilot engine as far as I can see. I agree it would probably be safer to use some sort of a "hex concat" function, but that requires an external library, idk if we have established use of such a thing yet. (afaict javascript does not have a suitable hex concatenation function; it always assumes the 0x doesn't exist)

Author: linked to elsewhere, tis well docced in this repo already 42ac22b

Reviewer: [Continuation](https://github.com/cowprotocol/docs/pull/595#discussion_r2930271284).

    let wrapperData = '0x';

    for (const [index, wrapper] of wrappers.entries()) {
        // Skip first wrapper's address (it's the transaction target)
        if (index !== 0) {
            // Add wrapper address (20 bytes, without 0x prefix)
            wrapperData += wrapper.target.slice(2);
        }

        // Encode data length as u16 big-endian (2 bytes)
        const dataLength = (wrapper.data.length - 2) / 2; // Remove '0x' and convert to byte length
        const lengthHex = dataLength.toString(16).padStart(4, '0');
        wrapperData += lengthHex;

        // Add wrapper data (without 0x prefix)
        wrapperData += wrapper.data.slice(2);
    }

    return wrapperData;
}
```

**Encoding Structure:**
```
For 2 wrappers:
[length₁(2B)][data₁][address₂(20B)][length₂(2B)][data₂]

For 3 wrappers:
[length₁(2B)][data₁][address₂(20B)][length₂(2B)][data₂][address₃(20B)][length₃(2B)][data₃]
```

##### 4. Call wrappedSettle

```typescript
const tx = {

>     to: wrappers[0].target,

Reviewer: It's nice to point out that the first target is missing in the encoding because it appears here.

Author: pointed out 42ac22b

    data: wrapperContract.interface.encodeFunctionData("wrappedSettle", [
        settleData,      // From step 2
        wrapperData      // From step 3
    ])
}
```

#### Using CowWrapperHelper

Alternatively, use the `CowWrapperHelper` contract for encoding and validation:

```solidity
// On-chain encoding and validation
address[] memory wrapperAddresses = [wrapper1, wrapper2];
bytes[] memory wrapperDatas = [data1, data2];

bytes memory wrapperData = CowWrapperHelper(HELPER).verifyAndBuildWrapperData(
    wrapperAddresses,
    wrapperDatas
);
```

In addition to producing the correct encoded data, the helper also validates:
- All wrappers are allowlisted
- All wrappers use the same settlement contract

> - Each wrapper can parse its data

Reviewer: I've seen this appearing before, possibly verbatim. We should just add a link, in any case the comments from before apply.

Author: [8d1e9a0](https://github.com/cowprotocol/docs/pull/595/commits/8d1e9a017526aa392008d9b4d26ca97da0e59b45)
